src/Dma.cpp

```diff
@@ -669,6 +669,20 @@ TimeSeries::_sp Dma::boundaryDemand() {
 }
 
 int Dma::allocateDemandToJunctions(time_t time) {
+  if (demandAllocationDelegate) {
+    auto response = demandAllocationDelegate->allocateDemands(time, *this);
+    switch (response) {
+      // success
+      case 0:
+      // error occurred
+      case 1:
+        return response;
+      // continue to standard demand allocation
+      case 2:
```

Comment:
Looking at the original allocate function, it returns 0 if success and 1 if failure. If the delegate follows this pattern why are we even switching on the response? Shouldn't the delegates allocation function just immediately be returned?

Reply:
because we want to also let the function just be ignored and move onto the code that happens after this switch statement

Reply:
So the allocate demand function returns an int and that can be any int, yet the switch statement only handles 0,1,2. And its also weird that the original function returns 0 or 1 only yet we are handling 2 here.

Reply:
that's because 2 is a new situation that we need to account for

Reply:
yep sounds good

Reply:
I understand that this particular RTX method has an integer 0/1 return pattern borrowed from epanet, but we don't have to inherit that silliness in the delegate.

```
if (this->_allocationDelegate && 
    this->_allocationDelegate->allocateDemands(share_me(this), time)) {
    return 0;  // delegate succeeded. no error and early out.
}
```

Reply:
oh wow there's a share_me

Reply:
Right that is what exceptions are for, but the way the old allocateDemandToJunctions function works is an error is just logged and then the simulation continues. It doesn't throw an exception, so I was doing the same.

Ok I'll just make it throw an exception and return a bool

Reply:
@sglindme great. the idea here is just to not have to subsume extra complexity within this part of the system. in client-supplied allocation algos, there may be a wide range of behavior that we want to handle - no reason to expose any of that here. The business logic in this code is meant to represent a baseline acceptable approach to demand allocation - naive and respecting the conservation of volume. client-supplied allocation delegates should be expected to have very specific failure modes which can be handled by client code.